src/ir/type-updating.cpp

```diff
@@ -385,21 +387,48 @@ void handleNonDefaultableLocals(Function* func, Module& wasm) {
 }
 
 Type getValidLocalType(Type type, FeatureSet features) {
-  // TODO: this should handle tuples with a non-nullable item
-  assert(canHandleAsLocal(type));
-  if (type.isNonNullable() && !features.hasGCNNLocals()) {
-    type = Type(type.getHeapType(), Nullable);
+  assert(type.isConcrete());
+  if (features.hasGCNNLocals()) {
+    return type;
+  }
+  if (type.isNonNullable()) {
+    return Type(type.getHeapType(), Nullable);
+  }
+  if (type.isTuple()) {
+    std::vector<Type> elems(type.size());
+    for (size_t i = 0, size = type.size(); i < size; ++i) {
+      elems[i] = getValidLocalType(type[i], features);
+    }
+    return Type(std::move(elems));
   }
   return type;
 }
 
 Expression* fixLocalGet(LocalGet* get, Module& wasm) {
-  if (get->type.isNonNullable() && !wasm.features.hasGCNNLocals()) {
+  if (wasm.features.hasGCNNLocals()) {
+    return get;
+  }
+  if (get->type.isNonNullable()) {
     // The get should now return a nullable value, and a ref.as_non_null
     // fixes that up.
     get->type = getValidLocalType(get->type, wasm.features);
     return Builder(wasm).makeRefAs(RefAsNonNull, get);
   }
+  if (get->type.isTuple()) {
+    auto type = get->type;
+    get->type = getValidLocalType(type, wasm.features);
+    std::vector<Expression*> elems(type.size());
+    Builder builder(wasm);
+    for (Index i = 0, size = type.size(); i < size; ++i) {
+      auto* elemGet =
+        i == 0 ? get : builder.makeLocalGet(get->index, get->type);
```

Comment:
I'm confused.  If the tuple has 3 elements, doesn't this look treat 0 as the (fixed up) tuple, and then 1 and 2 as the 2nd and 3rd indexes? What about the 1st index in the tuple?

That is, the first item in this loop has tuple type while all others do not..?

Reply:
The original `get` and all the new `get`s have the same full tuple type. It's the tuple extract below that picks a single element out of it.

Reply:
Oh, right... sorry, I was very confused here.